memgpt/server/server.py

```diff
@@ -848,12 +850,28 @@ def _agent_state_to_config(self, agent_state: AgentState) -> dict:
     def list_agents(
         self,
         user_id: uuid.UUID,
-    ) -> dict:
+        after: Optional[uuid.UUID] = None,
+        before: Optional[uuid.UUID] = None,
+        limit: Optional[int] = 20,
+        order: Literal["asc", "desc"] = "desc",
+        sort_by: Literal["created_at", "last_run", "name"] = "created_at",
+    ) -> List[AgentStateModel]:
         """List all available agents to a user"""
+        # TODO support sort_by
+        if sort_by != "created_at":
+            raise ValueError(sort_by)
+
         if self.ms.get_user(user_id=user_id) is None:
             raise ValueError(f"User user_id={user_id} does not exist")
 
-        agents_states = self.ms.list_agents(user_id=user_id)
+        agents_states = self.ms.list_agents(user_id=user_id, after=after, before=before, limit=limit, order=order)
+        # For the UI listing panel, we need to add the following metadata:
```

Comment:
@sarahwooders do you have any pointers on the most efficient way to get this metadata to the client (for each of the bullet points)? obviously the commented out code below is terrible (loading agents into cache)

additionally, i'm assuming we should pass all of this data back inside of `AgentStateModel.metadata`, correct? just treat `.metadata` as a free space to put extra data the client needs? or is the intended use different?

Reply:
yeah, we should probably have a list of reserved metadata fields (e.g. `_memgpt_last_run_timestamp`) to avoid conflicts with user metadata -- you could even represent the additional data as a pydantic model (so its easier to tell what fields it contains later on) but merge it with the metadata field of the `AgentStateModel` when returning